### PR TITLE
Pytest verifies and log-in into vault

### PR DESCRIPTION
**Pull request by jyejare**
Pytest session now verifies the vault session first.

- If the session expired, it takes you to login, and once logged in, it continues to the next step
- If the session is running, it just continues
- This only works for local execution and not in CI as CI has AppRole credentials directly!

Pytest STDOUT:
```
# pytest -s --collect-only tests/foreman/api/test_product.py
2023-09-11 14:02:28 - robottelo - WARNING - Warning! Vault not logged in!
2023-09-11 14:02:28 - robottelo - WARNING - Warning! The browser is about to open for vault OIDC login, close the tab once the sign-in is done!
2023-08-28 17:55:47 - robottelo - WARNING - Dynaconf validation failed, continuing for the sake of unit tests
    certs.cert_file is required in env main
2023-08-28 17:55:47 - robottelo - WARNING - Dynaconf validation failed, continuing for the sake of unit tests
    certs.cert_file is required in env main
=========== test session starts ==============
platform darwin -- Python 3.11.1, pytest-7.4.0, pluggy-1.0.0
Mandatory Requirements Mismatch: tenacity==8.2.3 dynaconf[vault]==3.2.1
Optional Requirements Mismatch: redis==5.0.0 sphinx==7.2.3
To update mismatched requirements, run the pytest command with '--upgrade-required-reqs' OR '--upgrade-all-reqs' option.
.
.
.
.
```

### CODE REVIEW

**Comment by ogajduse** [comment]
What is the problem that this PR is trying to solve? What would be a typical use of this new plugin?

**Comment by jyejare** [comment]
@ogajduse This saves you from getting out of pytest session for vault-login if its not already logged in into the vault by building the vault login facilities in pytest session itself.

So without worrying if the vault session is expired or not do the process by yourself of vault-login (helps new bees / devs where they would be unaware of they had to to login into vault first), you can directly execute pytest session.

Updated the description how the pytest session runs !

File: pytest_plugins/auto_vault.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+from pathlib import Path
+
+
+def pytest_addoption(parser):
+    """Options to allow user to update the requirements"""
+    envpath = Path('.env')
+    # check if this is being executed by a user and not automation/CI
+    if (
+        re.search(r'\s*#.*VAULT_SECRET_ID_FOR_DYNACONF', envpath.read_text())
+        and 'VAULT_SECRET_ID_FOR_DYNACONF' not in os.environ
```
Comment (Gauravtalreja1) [inline pytest_plugins/auto_vault.py:14]: I think this wouldn't handle the case where this variables are already set in dotenv?

File: pytest_plugins/auto_vault.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    ):
+        vstatus = subprocess.run("make vault-status", shell=True, capture_output=True).returncode
+        if vstatus != 0:
+            print('Vault token is expired, Logging you in ...')
+            subprocess.run('make vault-login', shell=True, stdout=subprocess.PIPE)
```
Comment (Gauravtalreja1) [inline pytest_plugins/auto_vault.py:19]: just curious, when vault-login script opens up a browser to authenticate, then how this is going to handle it? and is it going to wait until its ready?

**Review by Gauravtalreja1 — COMMENTED** [review]
LGTM, just a few questions

File: pytest_plugins/auto_vault.py (hunk 2 above)
Comment (rplevka) [inline pytest_plugins/auto_vault.py:19]: Could you use `timeout` parameter here to make it safer?

File: pytest_plugins/auto_vault.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+def pytest_addoption(parser):
+    """Options to allow user to update the requirements"""
+    envpath = Path('.env')
+    # check if this is being executed by a user and not automation/CI
+    if (
+        re.search(r'\s*#.*VAULT_SECRET_ID_FOR_DYNACONF', envpath.read_text())
+        and 'VAULT_SECRET_ID_FOR_DYNACONF' not in os.environ
+    ):
+        vstatus = subprocess.run("make vault-status", shell=True, capture_output=True).returncode
```
Comment (ogajduse) [inline pytest_plugins/auto_vault.py:16]: Could you stay inside Python?
Stack Python -> shell -> make -> Python. Please stay within Python.

**Review by ogajduse — CHANGES_REQUESTED** [review]

**Review by ogajduse — CHANGES_REQUESTED** [review]
Could you stay inside Python?
Stack Python -> shell -> make -> Python is a really ugly way to achieve what you want.

File: pytest_plugins/auto_vault.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+"""Plugin enables pytest to notify and update the requirements"""
+import os
+import re
+import subprocess
+from pathlib import Path
+
+
+def pytest_addoption(parser):
+    """Options to allow user to update the requirements"""
+    envpath = Path('.env')
```
Comment (ogajduse) [inline pytest_plugins/auto_vault.py:10]: If the user runs the tests from other than the root directory, pytest fails with:
```$ pwd
/home/ogajduse/repos/robottelo/tests/foreman/cli

$ pytest --collect-only test_hammer.py 
Traceback (most recent call last):
  File "/home/ogajduse/.virtualenvs/robottelo/bin/pytest", line 8, in <module>
    sys.exit(console_main())
             ^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 189, in console_main
    code = main()
           ^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 147, in main
    config = _prepareconfig(args, plugins)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 328, in _prepareconfig
    config = pluginmanager.hook.pytest_cmdline_parse(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_callers.py", line 55, in _multicall
    gen.send(outcome)
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/helpconfig.py", line 103, in pytest_cmdline_parse
    config: Config = outcome.get_result()
                     ^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_result.py", line 60, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 1075, in pytest_cmdline_parse
    self.parse(args)
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 1425, in parse
    self._preparse(args, addopts=addopts)
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 1327, in _preparse
    self.hook.pytest_load_initial_conftests(
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_callers.py", line 60, in _multicall
    return outcome.get_result()
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_result.py", line 60, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 1153, in pytest_load_initial_conftests
    self.pluginmanager._set_initial_conftests(
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 564, in _set_initial_conftests
    self._try_load_conftest(anchor, importmode, rootpath)
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 581, in _try_load_conftest
    self._getconftestmodules(anchor, importmode, rootpath)
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 610, in _getconftestmodules
    mod = self._importconftest(conftestpath, importmode, rootpath)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 658, in _importconftest
    self.consider_conftest(mod)
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 740, in consider_conftest
    self.register(conftestmodule, name=conftestmodule.__file__)
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 495, in register
    self.consider_module(plugin)
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 748, in consider_module
    self._import_plugin_specs(getattr(mod, "pytest_plugins", []))
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 755, in _import_plugin_specs
    self.import_plugin(import_spec)
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 792, in import_plugin
    self.register(mod, modname)
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/_pytest/config/__init__.py", line 488, in register
    ret: Optional[str] = super().register(plugin, name)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_manager.py", line 115, in register
    hook._maybe_apply_history(hookimpl)
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_hooks.py", line 300, in _maybe_apply_history
    res = self._hookexec(self.name, [method], kwargs, False)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_callers.py", line 60, in _multicall
    return outcome.get_result()
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_result.py", line 60, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/ogajduse/.virtualenvs/robottelo/lib/python3.11/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ogajduse/repos/robottelo/pytest_plugins/auto_vault.py", line 13, in pytest_addoption
    re.search(r'\s*#.*VAULT_SECRET_ID_FOR_DYNACONF', envpath.read_text())
                                                     ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/pathlib.py", line 1058, in read_text
    with self.open(mode='r', encoding=encoding, errors=errors) as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/pathlib.py", line 1044, in open
    return io.open(self, mode, buffering, encoding, errors, newline)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '.env'
```

Please find a way to always point to the correct location.

**Review by ogajduse — CHANGES_REQUESTED** [review]

File: pytest_plugins/auto_vault.py (hunk 2 above)
Comment (jyejare) [reply pytest_plugins/auto_vault.py:19]: The auth window will open and until we dont authenticate it holds, and finally after authentication it continues !

Comment (jyejare) [reply pytest_plugins/auto_vault.py:19]: I can add but this is step you have to authenticate if you are not logged into vault, you should not continue.

File: pytest_plugins/auto_vault.py (hunk 4 above)
Comment (jyejare) [reply pytest_plugins/auto_vault.py:10]: @ogajduse Yeah had that in mind but forgot to update the code. Thanks for catching !

**Comment by jyejare** [comment]
@ogajduse I Agree that it's an ugly way to achieve it but then functions in scripts are not made to be importable outside scripts. Hence same applies to `vault_login` script file. Since they cant we are calling them via `make` scripts.

We are wrapping on `vault` CLI utility in vault-login already so we are already following the path of python - Shell and had to repeat the same here with `python - make - python - vault`.

If we dont want to go this way, we had to externalize the vault function in utils and then use them in both scripting and this plugin.

File: pytest_plugins/auto_vault.py
Comment (ogajduse) [inline pytest_plugins/auto_vault.py:1]: Who requested this feature? I would like to see real people asking for this feature if we plan to merge it into our codebase.

**Comment by JacobCallahan** [comment]
Something to consider, if we're wanting to run this in every pytest session, is to bring the script logic into robottelo itself. This way we make the functionality here first-class and not offloaded to a script.

Alternatively, if we want to be less hand-holdy, we could log a warning/error informing the user that they need to run the vault-login make command.

File: pytest_plugins/auto_vault.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+
+def pytest_addoption(parser):
+    """Options to allow user to update the requirements"""
+    envpath = robottelo_root_dir.joinpath('.env')
+    # check if this is being executed by a user and not automation/CI
+    if (
+        re.search(r'\s*#.*VAULT_SECRET_ID_FOR_DYNACONF', envpath.read_text())
+        and 'VAULT_SECRET_ID_FOR_DYNACONF' not in os.environ
+    ):
```
[the remark this answers is not in the document]
Comment (jyejare) [reply pytest_plugins/auto_vault.py:16]: This checks if the pytest session is runing locally not in CI. In CI, the value `VAULT_SECRET_ID_FOR_DYNACONF` is set in env var and hence it does not need to login via OIDC, it logs in using AppRole credentials.

It is less likely the value is directly set in `.env` file from CI , but could be possible when one have app role creds and executes from local system. When this value is set , we should not run OIDC for vault login.

File: pytest_plugins/auto_vault.py
Comment (jyejare) [reply pytest_plugins/auto_vault.py:1]: There are no such ask but this is an enhancement in user experience. Also helps new bees / outside collaborators like DEVs to quickly execute the tests and no need to learn and worry about vault login first!

**Comment by ogajduse** [comment]
Output that I got from `make vault-login`:
```
$ make vault-login
[D 230914 13:58:52 __init__:31] Registered provider AnsibleTower
[D 230914 13:58:52 __init__:36] Registered action workflow for provider AnsibleTower
[D 230914 13:58:52 __init__:36] Registered action job_template for provider AnsibleTower
[D 230914 13:58:52 __init__:36] Registered action template for provider AnsibleTower
[D 230914 13:58:52 __init__:36] Registered action inventory for provider AnsibleTower
[D 230914 13:58:52 __init__:31] Registered provider Container
[D 230914 13:58:52 __init__:36] Registered action container_host for provider Container
[D 230914 13:58:52 __init__:36] Registered action container_app for provider Container
[D 230914 13:58:52 __init__:31] Registered provider TestProvider
[D 230914 13:58:52 __init__:36] Registered action test_action for provider TestProvider
Error looking up token: Error making API request.

URL: GET https://our.vault.server.com:8200/v1/auth/token/lookup-self
Code: 403. Errors:

* permission denied
2023-09-14 13:58:53 - robottelo - WARNING - Warning! The browser is about to open for vault OIDC login, close the tab once the sign-in is done!
Complete the login via your OIDC provider. Launching browser to:

    https://some.auth.server.in.our.internal.network.com/foo


Waiting for OIDC authentication to complete...
Success! You are now authenticated. The token information displayed below
is already stored in the token helper. You do NOT need to run "vault login"
again. Future Vault requests will automatically use this token.

Key                  Value
---                  -----
token                nopeiamnotgoingtorevealit
token_accessor       iamnotgoingtorevealthisonetoo
token_duration       20m
token_renewable      true
token_policies       ["default"]
identity_policies    ["app-foo-bar"]
policies             ["app-foo-bar" "default"]
token_meta_role      default
Key                  Value
---                  -----
token                nopeiamnotgoingtorevealit
token_accessor       iamnotgoingtorevealthisonetoo
token_duration       10h
token_renewable      true
token_policies       ["default"]
identity_policies    ["app-ro-mna-satqe-hashicorp"]
policies             ["app-ro-mna-satqe-hashicorp" "default"]
token_meta_role      default
Exception ignored in: <function Vault.__del__ at 0x7f0c99f57740>
Traceback (most recent call last):
  File "/home/ogajduse/repos/robottelo/robottelo/utils/vault.py", line 109, in __del__
AttributeError: 'NoneType' object has no attribute 'environ'
```

Successfully loaded the settings object:
```
$ python -c 'from robottelo.config import settings'
[D 230914 13:59:19 __init__:31] Registered provider AnsibleTower
[D 230914 13:59:19 __init__:36] Registered action workflow for provider AnsibleTower
[D 230914 13:59:19 __init__:36] Registered action job_template for provider AnsibleTower
[D 230914 13:59:19 __init__:36] Registered action template for provider AnsibleTower
[D 230914 13:59:19 __init__:36] Registered action inventory for provider AnsibleTower
[D 230914 13:59:19 __init__:31] Registered provider Container
[D 230914 13:59:19 __init__:36] Registered action container_host for provider Container
[D 230914 13:59:19 __init__:36] Registered action container_app for provider Container
[D 230914 13:59:19 __init__:31] Registered provider TestProvider
[D 230914 13:59:19 __init__:36] Registered action test_action for provider TestProvider
2023-09-14 13:59:24 - robottelo - WARNING - Dynaconf validation failed, continuing for the sake of unit tests
    vlan_networking.subnet is required in env main
2023-09-14 13:59:24 - robottelo - WARNING - Dynaconf validation failed, continuing for the sake of unit tests
    vlan_networking.subnet is required in env main
```

And from `make vault-logout` I got:
```
$ make vault-logout
[D 230914 14:01:04 __init__:31] Registered provider AnsibleTower
[D 230914 14:01:04 __init__:36] Registered action workflow for provider AnsibleTower
[D 230914 14:01:04 __init__:36] Registered action job_template for provider AnsibleTower
[D 230914 14:01:04 __init__:36] Registered action template for provider AnsibleTower
[D 230914 14:01:04 __init__:36] Registered action inventory for provider AnsibleTower
[D 230914 14:01:04 __init__:31] Registered provider Container
[D 230914 14:01:04 __init__:36] Registered action container_host for provider Container
[D 230914 14:01:04 __init__:36] Registered action container_app for provider Container
[D 230914 14:01:04 __init__:31] Registered provider TestProvider
[D 230914 14:01:04 __init__:36] Registered action test_action for provider TestProvider
Success! Revoked token (if it existed)
Exception ignored in: <function Vault.__del__ at 0x7f031f83b740>
Traceback (most recent call last):
  File "/home/ogajduse/repos/robottelo/robottelo/utils/vault.py", line 109, in __del__
AttributeError: 'NoneType' object has no attribute 'environ'
```

I must say that I liked the current version more as this new version spits out too much text that the users do not really need to see when they are "just logging in".

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: tests/foreman/api/test_product.py

File: pytest_plugins/auto_vault.py (hunk 1 above)
Comment (jyejare) [reply pytest_plugins/auto_vault.py]: @Gauravtalreja1 The condition is now changed to just verify if `VAULT_SECRET_ID_FOR_DYNACONF` environment var is set as thats the feasible check as per our CI, Local execution and PR checks!

**Comment by jyejare** [comment]
@ogajduse Thanks for verifying it locally on your system. Helps great!!

Though it looks good, I would like to :
1. Hide broker logs from vault output as people wont feel its related.
2. Fix the destructor exception!

**Comment by JacobCallahan** [comment]
@jyejare is this ready or should it be in draft?

**Comment by jyejare** [comment]
@JacobCallahan I think I can quickly fix the destructor exception in this PR and hiding logs for broker could be done in a separate follow up PR.

I haven't identified at how do I hide the broker logs yet , and even that should not block us from using this utility ! 

Rest of the testing looks good already including PRT, local testing, Make testing by @ogajduse etc.

**Comment by JacobCallahan** [comment]
@jyejare one simple option to try is copying the current environment variables, settings broker's log level to info/warn and passing that to subprocess.run via the env argument

**Comment by jyejare** [comment]
@JacobCallahan Got you about setting the broker log level to info / warn but did not get you on copying env vars and passing those envs to subprocess.run env argument. What would be the reason to do that ?

**Comment by jyejare** [comment]
@JacobCallahan As per our last discussion we should be good to merge this now ?

**Comment by jyejare** [comment]
@lhellebr PR is raised  #12822 to address the issue !

**Pull request merged**